src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt

```diff
@@ -148,50 +148,72 @@ class AwsS3BuildCacheService internal constructor(
 
     private fun Stopwatch.loadInternal(key: BuildCacheKey, reader: BuildCacheEntryReader): Boolean {
         val bucketPath = key.getBucketPath()
+        logger.info("Loading cache entry '{}' from S3 bucket", bucketPath)
+        val s3Object: ResponseBytes<GetObjectResponse>
         try {
-            logger.info("Loading cache entry '{}' from S3 bucket", bucketPath)
-            s3.getObject(bucketName, bucketPath).use { s3Object ->
-                val contentLength = s3Object.objectMetadata.contentLength
-                if (contentLength > maximumCachedObjectLength) {
+            s3Object = s3.getObject(GetObjectRequest.builder().bucket(bucketName).key(bucketPath).build(), AsyncResponseTransformer.toBytes())
+                    .join()
```

Comment:
Unfortunately, this has two issues:

1) It does receive and buffer the full response before the comparison of `contentLength > maximumCachedObjectLength`. That is it downloads full response instead of rejecting too large responses.
2) It does buffer the full blob to a byte array which might cause `OutOfMemoryError` in case the entry is big. Previous implementation streamed the response to the Gradle stream.

Reply:
Sorry for not noticing this! I have pushed a proposal commit, but it adds a ton of complexity to interface `Input/OutputStream` with the netty-based client's async API. I'm ok with reverting this commit and switching to the apache-based sync client, which presents `Input/OutputStream` natively, getting netty's efficient buffers would have been nice but not a dealbreaker. Let me know what you think.

FWIW, I have written an interface on the `putObject` side as well so that is now streamed appropriately. Maybe I can remove the reflection-based `file` optimization with it if we do go ahead and proceed with this change.

Reply:
reflection-based `file` optimization helps a lot: internally Gradle prepares the cache entry to a file on a local filesystem.
It makes it much more efficient to send that file to S3 as AWS API can retry the upload, and it can even split the file in chunks and upload it in parallel.

Reply:
I believe `OutputStream` is a wrong abstraction for cache upload:
1) It does not allow retries
2) It does not allow for sending the contents in chunks

I believe one day we could get `file()` as a public API from Gradle :)

Reply:
Ah thanks - that makes sense. So I guess rather than the `OutputStream` bridge I'll revert that to the memory copy approach since it's not used much in practice anyways.

For load, any suggestion on whether to stick with this async netty API, bridging to `InputStream` or sticking with sync apache API? I'm ok with either, the bridge is not so bad but pretty complex.